server.c

```diff
@@ -0,0 +1,211 @@
+
+		client.thread = client_thread;
+		clients[clientCounter] = client;
+		clientCounter += 1;
+
+		printf("Client connected\n");
+		printf("Waiting for client\n");
+	}
+
+	if (socket_client < 0)
+	{
+		perror("Accept failed");
+		return 1;
+	}
+
+	return 0;
+}
+
+void *ClientHandler(void *cli)
+{
+	struct sockaddr_in data_addr, data_client_addr;
+	int data_socket, data_socket_cli;
+	int c = sizeof(struct sockaddr_in);
+
+	Client client = *(Client *)cli;
+	//TODO: zastanowic sie nad rozmiarami buforow
+	char client_request[BUFSIZ], client_response[BUFSIZ];
+
+	//po otwarciu połączenia wysyłamy do klienta powitanie
+	sendResponse(client.socket, "220 (Welcome to STM32F4 FTP Server!) \r\n");
+
+	char *request_token;
+	//w petli odbieramy requesty od klienta
+	while (1)
+	{
+		recv(client.socket, client_request, BUFSIZ, 0);
+		// analizujemy request token po tokenie
+		// odcinamy pierwszą część (RETR, PASV, USER itp.)
+		request_token = strtok(client_request, " \r\n");
+		printf("%s\n", client_request);
+		// obsluga komendy USER
+		if (strcmp(request_token, "USER") == 0)
+		{
+			request_token = strtok(NULL, " \r\n");
+			if (request_token != NULL)
+			{
+				printf("%s\n", request_token);
+				strcpy(client.name, request_token);
+				sprintf(client_response, "230 (Hi %s!) \r\n", request_token);
+				sendResponse(client.socket, client_response);
+			}
+		}
+		// obsluga komendy PASV
+		if (strcmp(request_token, "PASV") == 0)
+		{
+			//trzeba otworzyc socket na porcie randomowym od 10k do 11k
+			data_socket = socket(AF_INET, SOCK_STREAM, 0);
+			data_addr.sin_family = AF_INET;
+			inet_pton(AF_INET, "127.0.0.1", &(data_addr.sin_addr));
+			//bierzemy DATA_PORT a potem inkrementujemy o jeden
+			do
+			{
+				data_addr.sin_port = htons(DATA_PORT);
+				DATA_PORT += 1;
+			} while (bind(data_socket, (struct sockaddr *)&data_addr, sizeof(data_addr)) < 0);
+
+			//wyciagamy z data_addr poszczegolne bajty
+			uint32_t ip_addr = ntohl(data_addr.sin_addr.s_addr);
+			uint16_t port = ntohs(data_addr.sin_port);
+			uint8_t b1 = (ip_addr >> 24), b2 = (ip_addr >> 16) % 256, b3 = (ip_addr >> 8) % 256, b4 = ip_addr % 256, b5 = port >> 8, b6 = port % 256;
+			sprintf(client_response, "227 Entering Passive Mode (%d,%d,%d,%d,%d,%d).\r\n", b1, b2, b3, b4, b5, b6);
+			sendResponse(client.socket, client_response);
+
+			//czekamy na połaczenie na portcie danych
+			listen(data_socket, 3);
+			printf("Waiting for data connection \n");
+			client.data_socket = accept(data_socket, (struct sockaddr *)&data_client_addr, (socklen_t *)&c);
+			printf("Data connection established\n");
+		}
+
+		// obsluga komendy QUIT
+		if (strcmp(request_token, "QUIT") == 0)
+		{
+			sprintf(client_response, "221 (Goodbye %s!) \r\n", client.name);
+			sendResponse(client.socket, client_response);
+			close(client.socket);
+
+			int *res = malloc(1 * sizeof(int));
+			*res = 0;
+			return (void *)res;
+		}
+
+		//obsluga komendy LIST
+		if (strcmp(request_token, "LIST") == 0)
```

Comment:
Serwer po zakończeniu wypisywania odpowiedzi do `client.data_socket` powinien tego socketa zamknąć. 
Możesz też domyślnie w client.data_socket wpisać jakieś -1 i dopisać do tego warunku `if (strcmp(request_token, "LIST") == 0)` dopisać sprawdzenie czy  `client.data_socket > 0`. Jeżeli wywołamy LIST a socket od danych klienta nie jest otwarty to serwer powinien zwrócić:
![image](https://user-images.githubusercontent.com/27630111/98105721-a199b700-1e98-11eb-9791-a074c104a0e5.png)